src/bindings/python/flux/cli/plugin.py

```diff
@@ -0,0 +1,187 @@
+                empty string if the plugin was invoked without VALUE.
+        """
+        pass
+
+    def modify_jobspec(self, args, jobspec, values):
+        """Allow plugin to modify jobspec
+
+        This function is called after arguments have been parsed and jobspec
+        has mostly been initialized.
+
+        Args:
+            args (:py:obj:`Namespace`): Namespace result from
+                :py:meth:`Argparse.ArgumentParser.parse_args()`.
+            jobspec (:obj:`flux.job.Jobspec`): instantiated jobspec object.
+                This plugin can modify this object directly to enact
+                changes in the current programs generated jobspec.
+            values (:obj:`Dict`): Dictionary of KEY[=VALUE] pairs returned
+                by parsing the arguments provided to the -P/--plugin
+                option. KEY is None if plugin was not invoked. VALUE is the
+                empty string if the plugin was invoked without VALUE.
+        """
+        pass
+
+    def validate(self, jobspec):
+        """Allow a plugin to validate jobspec
+
+        This callback may be used by the cli itself or a job validator
+        to validate the final jobspec.
+
+        On an invalid jobspec, this callback should raise ValueError
+        with a useful error message.
+
+        Args:
+            jobspec (:obj:`flux.job.Jobspec`): jobspec object to validate.
+        """
+        pass
+
+
+class CLIPluginRegistry:
+    """Flux CLI plugin registry helper class"""
+
+    def __init__(self):
+        self.plugins = []
+        etc = conf_builtin_get("confdir")
+        if etc is None:
+            raise ValueError("failed to get builtin confdir")
+        if getenv("FLUX_CLI_PLUGINPATH"):
+            self.plugindir = getenv("FLUX_CLI_PLUGINPATH")
+        else:
+            self.plugindir = f"{etc}/cli/plugins"
+
+    def _add_plugins(self, module, program):
+        entries = [
+            getattr(module, attr) for attr in dir(module) if not attr.startswith("_")
+        ]
+        ## TODO: as we get each entry, make sure it is NOT speciifically the CLIPlugin base class
+        for entry in entries:
+            # only process entries that are an instance of type (i.e. a class)
+            # and are a subclass of CLIPlugin:
+            if isinstance(entry, type) and issubclass(entry, CLIPlugin):
+                self.plugins.append(entry(program))
+
+    def load_plugins(self, program):
+        """Load all cli plugins from the standard path"""
+        for path in glob.glob(f"{self.plugindir}/*.py"):
+            self._add_plugins(import_path(path), program)
+        return self
+
+    def add_plugin_option(self, parser):
+        """Add the -P option and list of available keys"""
+        parser.add_argument(
+            "-P",
+            "--plugin",
+            type=CLIPluginValue,
+            action="append",
+            help=f"{self.get_plugin_help_messages()}",
```

Comment:
As you've noticed, this isn't going to display well in `--help` output.

Let's add a function to organize all the current plugin names and short usage messages into a nice table, then emit that in a custom section of the help output for the commands, e.g.:

```
Options provided by plugins:
    foo=VAL      Set option foo to VAL.
    bar[=VAL]    Set a bar with optional VAL.
```

Also, the `add_argument()` should not go in this class if we're treating them this way. Instead it just be with all the other `add_argument()` calls in the cli base.py.